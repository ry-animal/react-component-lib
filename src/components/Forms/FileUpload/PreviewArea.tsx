import * as React from "react";
import { useTheme } from "styled-components";
import { DefaultFile } from ".";
import IconClear from "../../../components/Icons/IconClear";
import { ButtonVariant, TextButton } from "../../Buttons/TextButton";
import { FilePreview } from "./FilePreview";
import STRINGS from "./index.strings";
import * as Styled from "./index.styled";
import { ISelectedFile } from "./ISelectedFile";

const BUTTON_LABEL = STRINGS.selectMedia;

type Props = {
  defaultValue?: DefaultFile;
  file?: ISelectedFile;
  hiddenFileInput: React.RefObject<HTMLInputElement>;
  maxFileSizeInMB: number;
  otherRestrictions?: string;
  onClose: () => void;
  blur?: boolean;
};

export function PreviewArea({
  defaultValue,
  file,
  hiddenFileInput,
  maxFileSizeInMB,
  otherRestrictions,
  onClose,
  blur,
}: Props) {
  const { color } = useTheme();

  return (
    <>
      {file || defaultValue ? (
        <Styled.FilePreviewWrapper>
          <Styled.ButtonWrapper>
            <Styled.DiscardButton
              aria-label={STRINGS.discard}
              onClick={onClose}
            >
              <IconClear fill={color.TEXT_1} />
            </Styled.DiscardButton>
          </Styled.ButtonWrapper>
          <FilePreview defaultValue={defaultValue} file={file} blur={blur} />
        </Styled.FilePreviewWrapper>
      ) : (
        <>
          <Styled.Instructions>
            {STRINGS.dragText(BUTTON_LABEL)}
          </Styled.Instructions>
          <TextButton
            label={BUTTON_LABEL}
            onClick={() => hiddenFileInput.current?.click()}
            variant={ButtonVariant.secondary}
            size="small"
          />
          <Styled.Restrictions>
            {STRINGS.maxFile(maxFileSizeInMB)}
          </Styled.Restrictions>
          {otherRestrictions && (
            <Styled.Restrictions>{otherRestrictions}</Styled.Restrictions>
          )}
        </>
      )}
    </>
  );
}
