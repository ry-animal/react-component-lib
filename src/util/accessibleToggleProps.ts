export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
