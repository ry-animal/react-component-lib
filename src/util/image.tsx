import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
