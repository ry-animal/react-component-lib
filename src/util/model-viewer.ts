/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
