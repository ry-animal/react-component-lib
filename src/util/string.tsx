const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
