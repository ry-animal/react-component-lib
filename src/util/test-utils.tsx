import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
