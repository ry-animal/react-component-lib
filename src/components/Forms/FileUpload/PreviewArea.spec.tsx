import userEvent from "@testing-library/user-event";
import { createRef } from "react";
import { render, screen } from "../../../util/test-utils";
import { PreviewArea } from "./PreviewArea";

describe("<PreviewArea />", () => {
  global.URL.createObjectURL = jest.fn(
    () => "blob:http://localhost:6006/af2e50bd-bc94-4dff-ae8c-440eabdac5c6"
  );
  const sampleProps = (fileSelected: boolean) => ({
    file: fileSelected
      ? {
          arrayBuffer: Buffer.from("Hello world!", "ascii").buffer,
          file: new File(["hello"], "world"),
          mimeType: "text/plain",
          name: "message.txt",
        }
      : undefined,
    hiddenFileInput: createRef<HTMLInputElement>(),
    maxFileSizeInMB: 1024,
    onClose: jest.fn(),
  });

  describe("rendering", () => {
    test("Default rendering matches snapshot (file selected)", () => {
      expect(
        render(<PreviewArea {...sampleProps(true)} />).asFragment()
      ).toMatchSnapshot();
    });

    test("Default rendering matches snapshot (file not selected)", () => {
      expect(
        render(<PreviewArea {...sampleProps(false)} />).asFragment()
      ).toMatchSnapshot();
    });

    test("diplay Other Restrictions", () => {
      render(
        <PreviewArea {...sampleProps(false)} otherRestrictions="Hello Jest" />
      );
      expect(screen.getByText("Hello Jest")).toBeInTheDocument();
    });
  });

  describe("event handling", () => {
    test("onChange called", () => {
      const props = sampleProps(true);
      render(<PreviewArea {...props} />);
      const button = screen.getByRole("button");
      userEvent.click(button);
      expect(props.onClose).toHaveBeenCalled();
    });
  });
});
