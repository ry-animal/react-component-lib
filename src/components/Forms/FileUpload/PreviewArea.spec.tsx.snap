// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PreviewArea /> rendering Default rendering matches snapshot (file not selected) 1`] = `
<DocumentFragment>
  .c0 {
  color: #404040;
  margin: 1em 30% 1em 30%;
}

<div
    class="c0"
  >
    Drag and drop your media files here or click the “Select Media” button below
  </div>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: unset;
  -ms-flex: unset;
  flex: unset;
  line-height: 1;
  gap: 8px;
}

.c0 {
  text-align: center;
  border: none;
  outline: none;
  overflow: hidden;
  cursor: pointer;
  pointer-events: auto;
  -webkit-transition: background-color 0.25s ease-out;
  transition: background-color 0.25s ease-out;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-height: 40px;
  font-weight: 600;
  padding: 0 28px;
  font-size: 16px;
  border-radius: 4px;
  background-color: transparent;
  box-shadow: inset 0 0 0 1px #000000;
  color: #000000;
}

.c0:disabled {
  opacity: 20%;
  cursor: default;
  pointer-events: none;
}

.c0:hover:not([disabled]),
.c0:focus-visible:not([disabled]) {
  box-shadow: inset 0 0 0 2px #000000;
}

<button
    aria-label="Select Media"
    class="c0"
  >
    <span
      class="c1"
    >
      Select Media
    </span>
  </button>
  .c0 {
  color: #A4A4A4;
  font-weight: 700;
  font-size: 0.8em;
  margin: 1em;
}

<div
    class="c0"
  >
    Max file size: 1024MB
  </div>
</DocumentFragment>
`;

exports[`<PreviewArea /> rendering Default rendering matches snapshot (file selected) 1`] = `
<DocumentFragment>
  .c4 {
  height: 150px;
  width: 150px;
}

.c5 {
  color: #404040;
}

.c3 {
  display: inline-block;
}

.c1 {
  position: absolute;
  right: 0;
  top: 0;
}

.c2 {
  background: none;
  border: 0;
  cursor: pointer;
}

.c0 {
  margin: 1em;
  position: relative;
}

<div
    class="c0"
  >
    <span
      class="c1"
    >
      <button
        aria-label="Discard"
        class="c2"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6L6 18"
            stroke="#151515"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M6 6L18 18"
            stroke="#151515"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </span>
    <div
      class="c3"
      data-testid="blur"
    >
      <img
        alt="A file has been selected"
        class="c4"
        src="[object Object]"
      />
    </div>
    <div
      class="c5"
    >
      message.txt
    </div>
  </div>
</DocumentFragment>
`;
